crl/validator.go

```diff
@@ -20,346 +20,420 @@ package crl
+		// TODO: the truststore should already contain ALL CAs that might be found in the verifiedChains
+		// TODO: should we check rawCerts, or verified chains?
+		// rawCerts contains raw cert data from the tls handshake.
+		// It is used together with the tls.Config to generate verifiedChains, but may contain additional certs that are not in a verified chain.
 
-	issuerName := crl.Issuer.String()
+		certificates, err := parseCertificatesSlice(rawCerts)
+		if err != nil {
+			return err
+		}
 
-	for _, cert := range crl.RevokedCertificates {
-		v.setRevoked(issuerName, cert.SerialNumber)
+		return v.validateChain(certificates)
 	}
-
-	v.lists[endpoint] = crl
-
 	return nil
 }
 
-// crlHasExpired is a replacement for pkix.CertificateList.HasExpired
-func crlHasExpired(crl *x509.RevocationList, deadline time.Time) bool {
-	return !deadline.Before(crl.NextUpdate)
+// validateChain validates that none of the certificates in the chain is revoked.
+// Certificated in chain are assumed ordered from leaf to root certificate.
+func (v *validator) validateChain(chain []*x509.Certificate) error {
+	var cert *x509.Certificate
+	var err error
+	for i := range chain {
+		cert = chain[len(chain)-1-i]
+		// check in reverse order to prevent CRL expiration errors due to revoked CAs no longer issuing CRLs
+		if err = v.validateCert(cert); err != nil {
+			return fmt.Errorf("%w: subject=%s, S/N=%s, issuer=%s", err, cert.Subject.String(), cert.SerialNumber.String(), cert.Issuer.String())
+		}
+	}
+	return nil
 }
 
-// IsSynced returns whether all the CRLs are downloaded and are not outdated (based on the offset)
-func (v *validator) IsSynced(maxOffsetDays int) bool {
-	endpoints := v.parseCRLEndpoints()
-
-	v.listsLock.RLock()
-	defer v.listsLock.RUnlock()
-
-	// Check if all CRLs have been downloaded, but if ignore CRL endpoints that are only used by expired certificates.
-	for endpoint, dependingCAs := range endpoints {
-		if !anyCertificateActive(dependingCAs) {
-			continue
+func (v *validator) validateCert(cert *x509.Certificate) error {
+	if nowFunc().After(cert.NotAfter) {
+		return ErrCertExpired
+	}
+	for _, endpoint := range cert.CRLDistributionPoints {
+		crl, ok := v.getCRL(endpoint)
+		if ok && crl.revoked[cert.SerialNumber.String()] != nil {
+			// revocation takes precedence over an internal error.
+			return ErrCertRevoked
+		}
+		if !ok || crl.lastUpdated.IsZero() {
+			// TODO: async or not? This method call is the result of a client request running in its own go routine. We can wait for the crl to be downloaded...
+			// Currently the request fails on the first request using a certificate with an unknown CRl endpoint
+			go v.updateCRL(endpoint, nil)
+			return ErrCRLMissing
 		}
-		// downloaded?
-		if _, ok := v.lists[endpoint]; !ok {
-			return false
+		if !nowFunc().Before(crl.list.NextUpdate) {
+			// CA expiration is checked earlier in the chain
```

Comment:
we had a configurable offset before (max time CRL is allowed to be out. of date)

Reply:
I consider the offset to be a bug. The CRLs have an expiration time (`NextUpdate`), and are updated long before they expire. We needed the offset because the current validator does not download the new one until the previous one had expired and was technically invalid. This PR fixes that, so the offset is no longer needed.

Reply:
If we remove the option, remember to add it to the release notes